streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -342,7 +342,18 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         maybeThrowTaskExceptions(taskCloseExceptions);
 
-        createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        final Collection<Task> newActiveTasks = createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        // If there are any transactions in flight and there are newly created active tasks, commit the tasks
+        // to avoid potential long restoration times.
+        if (processingMode == EXACTLY_ONCE_V2 && threadProducer().transactionInFlight() && !newActiveTasks.isEmpty()) {
+            log.info("New active tasks were added and there is an inflight transaction. Attempting to commit tasks.");
+            final int numCommitted = commitTasksAndMaybeUpdateCommittableOffsets(newActiveTasks, new HashMap<>());
+            if (numCommitted == -1) {
```

Comment:
This reminds me one thing: we call `onAssignment` first, and then `onPartitionsAssigned` later, and we only set `rebalanceInProgress` to false in the latter func, which means that during `onAssignment` we would always see `rebalanceInProgress == true` which would not allow a commit logically.. 

I gave some thought about it, and currently the quick (and somewhat dirty..) fix would be to move the `rebalanceInProgress == true` line right before the `createNewTasks` inside the `handleAssignment` here. But we should leave a TODO such that moving one we would only rely on `onAssignment` as the rebalance complete barrier and move others from `onPartitionsAssigned` as well.

Reply:
Also cc @lucasbru @mjsax @lihaosky  to bring to your attention.

Reply:
Thanks for this comment. IIUC, the suggestion is to move the `rebalanceInProgress == true` from `commitTasksAndMaybeUpdateCommittableOffsets` before creating new tasks and not create any new tasks or try to commit them if there's a rebalance in progress? 
Also, does the check in `commitTasksAndMaybeUpdateCommittableOffsets` still remain? I see it is being called from a couple of other places as well like `handleCorruption` and `commit`.

Reply:
Sorry I made a typo there, I meant that we move the `rebalanceInProgress = false` line, which is currently in `onPartitionsAssigned` -> `onRebalanceComplete` into `onAssignment`, right before the `createNewTasks` line. By doing so we still keep the checking, knowing that at the time when we reach `onAssignment` it means the rebalance is basically done, and the check inside `commitTasksAndMaybeUpdateCommittableOffsets` would not block us.

We would still have the check inside `commitTasksAndMaybeUpdateCommittableOffsets` for other callers.

Reply:
Thanks @guozhangwang for clarifying. I moved the `rebalance=false` assignment from `handleRebalanceComplete` into `handleAssignment` right before `createNewTasks`. It appears to me that it has a side effect. This test in `StreamThreadTest#shouldNotEnforceRebalanceWhenCurrentlyRebalancing` is a good indicator where in it's expecting a clean shutdown and a state transition to DEAD state but it's never happening.

Because the rebalanceInProgress is never falsified, the `runLoop` stays in an infinite loop in `StreamThread` since this condition is never met:

```
if (!taskManager.rebalanceInProgress() && nextProbingRebalanceMs.get() < time.milliseconds()) {
                    log.info("Triggering the followup rebalance scheduled for {}.", Utils.toLogDateTimeFormat(nextProbingRebalanceMs.get()));
                    mainConsumer.enforceRebalance("triggered followup rebalance scheduled for " + nextProbingRebalanceMs.get());
                    nextProbingRebalanceMs.set(Long.MAX_VALUE);
                }
```
In the successful case(i.e with the rebalanceInProgress=false; retained in `handleRebalanceComplete` I see this: 

```
[2023-04-05 14:39:20,664] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Informed to shut down (org.apache.kafka.streams.processor.internals.StreamThread:1168)
[2023-04-05 14:39:22,442] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] State transition from STARTING to PENDING_SHUTDOWN (org.apache.kafka.streams.processor.internals.StreamThread:239)
[2023-04-05 14:40:31,036] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Triggering the followup rebalance scheduled for 2023-04-05 14:39:14,762 +05:30. (org.apache.kafka.streams.processor.internals.StreamThread:621)
[2023-04-05 14:40:31,039] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Shutting down clean (org.apache.kafka.streams.processor.internals.StreamThread:1182)
[2023-04-05 14:40:31,060] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] State transition from PENDING_SHUTDOWN to DEAD (org.apache.kafka.streams.processor.internals.StreamThread:239)
[2023-04-05 14:40:31,060] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Shutdown complete (org.apache.kafka.streams.processor.internals.StreamThread:1217)
```

while in the other case, the follow up rebalance is never scheduled which means the current one never ends and the StreamThread can never shutdown looks like:

```
[2023-04-05 14:43:11,201] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Starting (org.apache.kafka.streams.processor.internals.StreamThread:572)
[2023-04-05 14:43:11,205] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] State transition from CREATED to STARTING (org.apache.kafka.streams.processor.internals.StreamThread:239)
[2023-04-05 14:43:11,221] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] Informed to shut down (org.apache.kafka.streams.processor.internals.StreamThread:1168)
[2023-04-05 14:43:11,221] INFO stream-thread [stream-thread-test-87bf53a8-54f2-485f-a4b6-acdbec0a8b3d-StreamThread-1] State transition from STARTING to PENDING_SHUTDOWN (org.apache.kafka.streams.processor.internals.StreamThread:239)
```

Reply:
Hmm.. that was not expected, did you check the trace and find out why the `onAssignment` was not triggered, in which it would set `rebalanceInProgress = true`?

Reply:
Yeah that test didn't trigger `onAssignment` and invoked `handleRebalanceComplete` directly which explains the problem which means that for that test to work, we need to retain the `rebalanceInProgress = false` in `handleRebalanceComplete` as well. I think it should be ok to have this line in both `handleAssignment` and `handleRebalanceComplete` because the latter anyways signifies a rebalance completion.

But, there are a couple of other things which seem to break now (sorry :(  ) because of this new condition. One of them is `StreamThread#shouldOnlyCompleteShutdownAfterRebalanceNotInProgress` . This one is interesting because it is invoking `handleAssignment` and then `start` and `stop` on the StreamThread and asserting that while rebalanance is in progress, active tasks are still returned. Now because, we are explicitly setting `rebalanceInProgress` to false in `handleAssignment` itself, this test fails. Since this is EOS-V1 related test, I moved the `rebalanceInProgress = false` inside the EOSv2 check and this test passes.

But then, because of `rebalanceInProgress = false` within `handleAssignment`, `EosV2UpgradeIntegrationTest#shouldUpgradeFromEosAlphaToEosV2` fails when clients are closed cleanly. This one passes the moment I remove the line `rebalanceInProgress = false` from `handleAssignment`.

All in all, this one line seems to be causing side effects at various places!

Reply:
Seems we should not mess with the control flow here. Me might end up with spagetti-code. I would advocate the move the commit logic into the restore code path if possible as mentioned further above -- this way, the existing control flow is not changed and we avoid all these compilations.

Reply:
I see. Either ways, moving the flag around is failing some or the other condition. I have asked Lucas in the above thread.